File: C/external.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <sodium.h>

#include "adrs.h"
#include "params.h"
#include "internal.h"
#include "shake.h"
#include "external.h"

// Algorithmus 21: Generiert ein SLH-DSA Schlüsselpaar
void slh_keygen(Parameters *prm, uint8_t *SK_seed, uint8_t *SK_prf, uint8_t *PK_seed, uint8_t *SK, uint8_t *PK)
{
    if (sodium_init() < 0) {
        printf("Error initalizing sodium library\n");
        return;
    }

    // check if the seed and prf values are empty
    // if so, randomly generate them
    uint32_t sum_sk_seed, sum_sk_prf, sum_pk_seed = 0;
    for (uint32_t i = 0; i < prm->n; i++) {
        sum_sk_seed |= SK_seed[i];
        sum_sk_prf  |= SK_prf[i];
        sum_pk_seed |= PK_seed[i];
    }
    if (sum_sk_seed == 0)
        randombytes_buf(SK_seed, prm->n);
    if (sum_sk_prf == 0)
        randombytes_buf(SK_prf , sizeof SK_prf);
    if (sum_pk_seed == 0)
        randombytes_buf(PK_seed, sizeof PK_seed);

    slh_keygen_internal(prm, SK_seed, SK_prf, PK_seed, SK, PK);
}

// Algorithmus 22: Generiert eine reine SLH-DSA Signatur
void slh_sign(Parameters *prm, uint8_t *M, size_t M_len, const uint8_t *ctx, const size_t ctx_len, const uint8_t *SK, uint8_t *SIG, bool deterministic)
{
    if (ctx_len > MAX_CTX_LENGTH) {
        printf("Invalid context length\n");
        return;
    }

    // for deterministic varaiant, use PK_seed for addrnd
    uint8_t addrnd[prm->n];
    if (deterministic == true) {
        memcpy(addrnd, SK + 2 * prm->n, prm->n);
    }
    else {
        if (sodium_init() < 0) {
            printf("Error initalizing sodium library\n");
            return;
        }
        randombytes_buf(addrnd, sizeof addrnd);
    }
    uint8_t M_prime[1 + 1 + ctx_len + M_len];
    M_prime[0] = 0;
    toByte(ctx_len, 1, M_prime + 1);

    memcpy(M_prime + 2, ctx, ctx_len);
    memcpy(M_prime + 2 + ctx_len, M, M_len);

    slh_sign_internal(prm, M_prime, sizeof M_prime, SK, addrnd, SIG);
}

// Algorithmus 23: Generiert eine vorgehashte SLH-DSA Signatur
void hash_slh_sign(Parameters *prm, const uint8_t *M, size_t M_len, const uint8_t *ctx, size_t ctx_len, const char *PH, const uint8_t *SK, uint8_t *SIG)
{
    if (ctx_len > MAX_CTX_LENGTH) {
        printf("Context is longer that %d\n", MAX_CTX_LENGTH);
        return;
    }

    uint8_t addrnd[prm->n];
    if (sodium_init() < 0) {
        printf("Error initalizing sodium library\n");
        return;
    }
    // randombytes_buf(addrnd, sizeof addrnd);
    memcpy(addrnd, SK + 2 * prm->n, prm->n);

    uint8_t OID[11];
    uint8_t PHM[64] = {0};

    if (strcmp(PH, "SHA-256") == 0) {
        memcpy(OID, "\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x01", 11);
        SHA_256(M, M_len, PHM);
    } else if (strcmp(PH, "SHA-512") == 0) {
        memcpy(OID, "\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x03", 11);
        SHA_512(M, M_len, PHM);
    } else if (strcmp(PH, "SHAKE128") == 0) {
        memcpy(OID, "\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x0B", 11);
        SHAKE_128(M, M_len, PHM, 32);
    } else if (strcmp(PH, "SHAKE256") == 0) {
        memcpy(OID, "\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x0C", 11);
        SHAKE_256(M, M_len, PHM, 64);
    } else {
        printf("Unsupported hash function\n");
        printf("Valid values are 'SHA-256', 'SHA-512', 'SHAKE128', 'SHAKE256'\n");
        return;
    }

    uint8_t M_prime[1 + 1 + ctx_len + 11 + sizeof PHM];
    M_prime[0] = 1;
    toByte(ctx_len, 1, M_prime + 1);

    memcpy(M_prime + 2, ctx, ctx_len);
    memcpy(M_prime + 2 + ctx_len, OID, sizeof OID);
    memcpy(M_prime + 2 + ctx_len + sizeof OID, PHM, sizeof PHM);

    slh_sign_internal(prm, M_prime, sizeof M_prime, SK, addrnd, SIG);
}

// Algorithmus 24: Verifiziert eine reine SLH-DSA Signatur
bool slh_verify(Parameters *prm, uint8_t *M, size_t M_len, uint8_t *SIG, size_t SIG_len, uint8_t *ctx, size_t ctx_len, const uint8_t *PK)
{
    if (ctx_len > MAX_CTX_LENGTH) {
        printf("Invalid context length\n");
        return false;
    }
    uint8_t M_prime[1 + 1 + ctx_len + M_len];
    M_prime[0] = 0;
    toByte(ctx_len, 1, M_prime + 1);

    memcpy(M_prime + 2, ctx, ctx_len);
    memcpy(M_prime + 2 + ctx_len, M, M_len);
    return slh_verify_internal(prm, M_prime, sizeof M_prime, SIG, SIG_len, PK);
}

// Algorithmus 25: Verifiziert eine vorgehashte SLH-DSA Signatur
bool hash_slh_verify(Parameters *prm, const uint8_t *M, size_t M_len, uint8_t *SIG, size_t SIG_len, const uint8_t *ctx, size_t ctx_len, const char *PH, const uint8_t *PK)
{
    if (ctx_len > MAX_CTX_LENGTH) {
        printf("Context is longer that %d\n", MAX_CTX_LENGTH);
        return false;
    }

    uint8_t OID[11];
    uint8_t PHM[64] = {0};

    if (strcmp(PH, "SHA-256") == 0) {
        memcpy(OID, "\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x01", 11);
        SHA_256(M, M_len, PHM);
    } else if (strcmp(PH, "SHA-512") == 0) {
        memcpy(OID, "\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x03", 11);
        SHA_512(M, M_len, PHM);
    } else if (strcmp(PH, "SHAKE128") == 0) {
        memcpy(OID, "\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x0B", 11);
        SHAKE_128(M, M_len, PHM, 32);
    } else if (strcmp(PH, "SHAKE256") == 0) {
        memcpy(OID, "\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x0C", 11);
        SHAKE_256(M, M_len, PHM, 64);
    } else {
        printf("Unsupported hash function\n");
        printf("Valid values are 'SHA-256', 'SHA-512', 'SHAKE128', 'SHAKE256'\n");
        return false;
    }

    uint8_t M_prime[1 + 1 + ctx_len + 11 + sizeof PHM];
    M_prime[0] = 1;
    toByte(ctx_len, 1, M_prime + 1);

    memcpy(M_prime + 2, ctx, ctx_len);
    memcpy(M_prime + 2 + ctx_len, OID, sizeof OID);
    memcpy(M_prime + 2 + ctx_len + sizeof OID, PHM, sizeof PHM);

    return slh_verify_internal(prm, M_prime, sizeof M_prime, SIG, SIG_len, PK);
}

File: python/external.py
import secrets
import hashlib

import params
from adrs import toByte
from internal import slh_keygen_internal, slh_sign_internal, slh_verify_internal

# Algorithmus 21 (Generates an SLH-DSA key pair)
def slh_keygen(SK_seed: bytes = b"", SK_prf: bytes = b"", PK_seed: bytes = b"") -> tuple:

    # Generate random seeds
    if not SK_seed:
        SK_seed = secrets.token_bytes(params.prm.n)
    if not SK_prf:
        SK_prf = secrets.token_bytes(params.prm.n)
    if not PK_seed:
        PK_seed = secrets.token_bytes(params.prm.n)

    # Check for errors
    if not SK_seed or not SK_prf or not PK_seed:
        return None, None

    # Call the internal key generation function
    return slh_keygen_internal(SK_seed, SK_prf, PK_seed)    # Output: SLH-DSA key pair (SK, PK)


# Algorithmus 22 (Generates a pure SLH-DSA signature)
def slh_sign(M: bytes, SK: bytes, deterministic: bool = True) -> bytes:

    # for deterministic variant, use PK_seed for addrnd
    if deterministic:
        addrnd = SK[2 * params.prm.n:3 * params.prm.n]
    else:
        addrnd = secrets.token_bytes(params.prm.n)
        if addrnd is None:
            return b""

    return slh_sign_internal(M, SK, addrnd)


# Algorithmus 23 (Generates a pre-hash SLH-DSA signature)
def hash_slh_sign(M: bytes, ctx: list, PH: str, SK: bytes, deterministic: bool = True) -> bytes:

    if len(ctx) > 255:
        return b""

    if deterministic:
        addrnd = SK[2 * params.prm.n:3 * params.prm.n]
    else:
        addrnd = secrets.token_bytes(params.prm.n)  # Skip for deterministic variant
        if addrnd is None:
            return b""

    # NOTE we initialize PHM as an array with 64 bytes
    PHM = bytearray([0] * 64)

    # Pre-hash the message
    match PH:
        case "SHA-256":
            OID = toByte(0x0609608648016503040201, 11)
            PHM[0:32] = hashlib.sha256(M).digest()
        case "SHA-512":
            OID = toByte(0x0609608648016503040203, 11)
            PHM[0:64] = hashlib.sha512(M).digest()
        case "SHAKE128":
            OID = toByte(0x060960864801650304020B, 11)
            PHM[0:32] = hashlib.shake_128(M).digest(32)
        case "SHAKE256":
            OID = toByte(0x060960864801650304020C, 11)
            PHM[0:64] = hashlib.shake_256(M).digest(64)
        case _:
            # Handle other approved hash functions or XOFs
            raise NotImplementedError("Unsupported pre-hash function")

    # Construct the M' message
    M_prime = toByte(1, 1) + toByte(len(ctx), 1) + ctx + OID
    M_prime = bytearray(M_prime) + PHM

    return slh_sign_internal(M_prime, SK, addrnd)  # Omit addrnd for deterministic variant


# Algorithmus 24 (Verifies a pure SLH-DSA signature)
def slh_verify(M: bytes, SIG: bytes, PK: bytes) -> bool:
    return slh_verify_internal(M, SIG, PK)


# Algorithmus 25 (Verifies a pre-hash SLH-DSA signature)
def hash_slh_verify(M: bytes, SIG: bytes, ctx: list, PH: str, PK: bytes) -> bool:   # Input: Message 𝑀, signature SIG, context string 𝑐𝑡𝑥, pre-hash function PH, public key PK

    if len(ctx) > 255:
        return False

    PHM = bytearray([0] * 64)
    # Pre-hash the message
    match PH:
        case "SHA-256":
            OID = toByte(0x0609608648016503040201, 11)
            PHM[0:32] = hashlib.sha256(M).digest()
        case "SHA-512":
            OID = toByte(0x0609608648016503040203, 11)
            PHM[0:64] = hashlib.sha512(M).digest()
        case "SHAKE128":
            OID = toByte(0x060960864801650304020B, 11)
            PHM[0:32] = hashlib.shake_128(M).digest(32)
        case "SHAKE256":
            OID = toByte(0x060960864801650304020C, 11)
            PHM[0:64] = hashlib.shake_256(M).digest(64)
        case _:
            # Handle other approved hash functions or XOFs
            raise NotImplementedError("Unsupported pre-hash function")

    # Construct the M' message
    M_prime = toByte(1, 1) + toByte(len(ctx), 1) + ctx + OID
    M_prime = bytearray(M_prime) + PHM

    return slh_verify_internal(M_prime, SIG, PK)
